New module: Add module to collect Amazon AWS IAM user info (cloud/amazon/iam_user_info)

psharkey opened the pull request:
##### SUMMARY
For consideration, `iam_user_info.py` is a new module for gathering AWS IAM user info. Based on https://github.com/ansible/ansible-modules-extras/pull/2314 with an additional group parameter. 

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
 iam_user_info.py

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.2.1.0
  config file = /home/ansible.cfg
  configured module search path = Default w/o overrides
```


##### ADDITIONAL INFORMATION
Example playbook:
```
---
- hosts: localhost
  gather_facts: false
  tasks:
  - name: get IAM users
    iam_user_info:
      group: fumpti
      name: schrunch
    register: users

  - debug: var=users
```
Debug output from example (using `stdout_callback = yaml` in ansible.cfg):

```
TASK [debug]  
ok: [localhost] => 
  users:
    changed: false
    failed: false
    iam_users:
    - arn: arn:aws:iam::<REDACTED>:user/schrunch
      create_date: '2017-02-01T19:18:47+00:00'
      password_last_used: '2019-09-20T00:49:54+00:00'
      path: /
      user_id: <REDACTED>
      user_name: schrunch
```

psharkey committed:
AWS: new module iam_user_info

Signed-off-by: psharkey <psharkey@cleo.com>
Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

Rename from iam_user_facts to iam_user_info.

Rename and target 2.10.

Fixing docs.

ryansb commented on lib/ansible/modules/cloud/amazon/iam_user_facts.py:
```diff
@@ -0,0 +1,181 @@
+    description:
+     - The group name name of the IAM user to look for. Mutually exclusive with C(path).
+    required: false
+  path:
```
Can you add an example or explanation of group vs. path, and when each would be used?

ryansb commented on lib/ansible/modules/cloud/amazon/iam_user_facts.py:
```diff
@@ -0,0 +1,181 @@
+        try:
+            iam_users = connection.list_users(**params)['Users']
+        except (ClientError, ParamValidationError) as e:
+            module.fail_json(msg=e.message, **camel_dict_to_snake_dict(e.response))
```
Can you also catch generic exception here and raise the traceback like `module.fail_json(msg="Failed while listing IAM users", exception=traceback.format_exc())`?

ryansb reviewed (changes requested):
Overall looks good.

Akasurde commented:
@psharkey Are you still working on this ? Let me know so that I can take over this PR.

psharkey commented:
@Akasurde You can take over this PR.

1161213 commented:
To be honest I don't know the right way to do this but I am just writing this comment because from what I can understand the "iam_user_facts" module is ready to be shipped but I cant find it anywhere.
Is it an error? Or am I missing something?

psharkey commented:
@1161213 Are you thinking of [iam_role_facts](https://docs.ansible.com/ansible/latest/modules/iam_role_facts_module.html#iam-role-facts-module)?

1161213 commented:
> @1161213 Are you thinking of [iam_role_facts](https://docs.ansible.com/ansible/latest/modules/iam_role_facts_module.html#iam-role-facts-module)?

No, what I really need is a module to list the users of IAM. Searching the Internet i found this module but I can't find it in the Ansible documentation and I was trying to understand if it is ready to be used or if will be ready at all

psharkey commented:
The [documentation](https://github.com/ansible/ansible/blob/00d0507b8318fc798bd8945c52b33a0d40bf2f26/lib/ansible/modules/cloud/amazon/iam_user_facts.py#L20-L102) is included in the source code. This PR isn't merged yet so this is the only place you'll find the documentation. 
To try this yourself you'd have to [add it to your local ansible](https://docs.ansible.com/ansible/latest/dev_guide/developing_locally.html#adding-a-module-locally) installation.

tremble commented:
@psharkey Sorry to do this after so long.

If you're still interested in getting this in it should be renamed as iam_user_info additionally is there any chance you could add some integration tests?

@willthames wrote some good documentation for getting things set up http://willthames.github.io/2017/07/17/so-you-want-to-test-aws-modules-for-ansible.html

A simple example can be found in test/integration/targets/aws_codecommit

psharkey commented:
@tremble Thanks for looking at this! Yes, I'd still like to get this in and will work on getting this updated and adding an integration test. 
Would `test/integration/targets/iam_group` be a good starting point for an integration test? I'd probably follow the same pattern for creating/removing  a test user and group. I'm guessing something will have to be done so this will be allowed to run on the CI server?

tremble commented:
iam_group should be a good start.

Since you'll need to manipulate users what's probably worth doing is creating an iam_user test suite and adding "iam_user_info" into the aliases file.  (Unfortunately one doesn't exist yet, I'm slowly working my way through some of them)  For example https://github.com/ansible/ansible/blob/devel/test/integration/targets/iam_group/aliases

The first part would be getting this working in your own environment.  You can find the relevant policies in https://github.com/ansible/ansible/tree/devel/hacking/aws_config/testing_policies

Once you've got them working we can talk to @jillr / @s-hertel about getting the updates made to the test accounts.  It's possible they'll not be added to the "shippable" test suites creating users has complex security implications (that's what 'unsupported' in aliases means)

psharkey committed:
Adding iam_user_info integration test.

Removing unnecessary tasks.

Fixing yamllint failure test/integration/targets/iam_user_info/defaults/main.yml:5:1: empty-lines: too many blank lines (1 > 0).

psharkey commented:
@tremble Can you please look at the integration tests added as ['unsupported'](https://github.com/ansible/ansible/blob/1e2f8eb5b881d48be4c9cdb1259f0e4008a9293d/test/integration/targets/iam_user_info/aliases#L2) (for now)?

tremble commented on test/integration/targets/iam_user_info/tasks/main.yml:
```diff
@@ -0,0 +1,105 @@
+  - name: assert correct keys for path are returned
+    assert:
+      that:
+        - iam_user_info.iam_users[0].arn is not none
```
niggle: You can do a quick test that this is actually an ARN:
```
....arn.startswith("arn:aws:iam")
....arn.endswith("user/{{ test_user }}")
```

psharkey replied:
https://github.com/ansible/ansible/pull/23382/commits/9f67467fdd9c5ab480bc1d8e0d2aa20f20599b7d adds some more detailed checks.

tremble commented on lib/ansible/modules/cloud/amazon/iam_user_info.py:
```diff
@@ -0,0 +1,189 @@
+        )
+    )
+
+    module = AnsibleModule(
```
This can be migrated to AnsibleAWSModule, which will handle things like the boto requirements and has a nice fail_json_aws method

psharkey replied:
https://github.com/ansible/ansible/pull/23382/commits/e8d91f4a221d6d196b5d48988f55bcc61764f46f and https://github.com/ansible/ansible/pull/23382/commits/df9d52172242e86b88497c7468e9a3d25c8bf6f4 have the related updates here.

tremble commented on test/integration/targets/iam_user_info/defaults/main.yml:
```diff
@@ -0,0 +1,4 @@
+---
+test_group: '{{ resource_prefix }}-group'
+test_path: '/'
```
If you're testing searching with paths it's worth testing with something like `/{{ resource_prefix }}/` too

psharkey replied:
Only the default path ('/') is possible here since [iam_user](https://docs.ansible.com/ansible/latest/modules/iam_user_module.html) doesn't support the [`Path`](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/iam.html#IAM.Client.create_user) parameter.

tremble commented on test/integration/targets/iam_user_info/aliases:
```diff
@@ -0,0 +1,2 @@
+cloud/aws
```
it would be good to move these tests to iam_user and add iam_user_info as an alias.

psharkey replied:
https://github.com/ansible/ansible/pull/23382/commits/f44a16ce370dcfa99846097ce0acefd9afd9bc3a moves these and adds iam_user_info as an alias.

tremble reviewed (changes requested):
Just tried running these tests using ansible-test and it failed with

```
botocore.exceptions.NoCredentialsError: Unable to locate credentials
```

I suspect there's something in the boto client initiation that you're not quite doing right.  Rather than creating the connection yourself you can also use the AnsibleAWSModule class and let it manage most of this for you.

```
argument_spec = dict( ..... ) # no need to add the boto pieces AnsibleAWSModule does this for you
module = AnsibleAWSModule(argument_spec=argument_spec, supports_check_mode=True)
connection = module.client('iam')
```

See iam_role for an example.

It would also be good if you could add the IAM policies attached to the user.  Again, iam_role_info has a nice example of this.  (See describe_iam_role)

psharkey committed:
name paramter is optional

psharkey committed:
Switch to use AnsibleAWSModule.

psharkey committed:
Convert to using fail_json_aws

psharkey committed:
Rework asserts to inspect ARN.

psharkey committed:
Move integration tests from iam_user_info to iam_user.

psharkey committed:
Fix pep8 problems.

psharkey commented:
Integration tests may need to be adjusted depending on the timing of https://github.com/ansible/ansible/pull/59079.

tremble commented on lib/ansible/modules/cloud/amazon/iam_user_info.py:
```diff
@@ -0,0 +1,184 @@
+
+
+def main():
+    argument_spec = ec2_argument_spec()
```
AnsibleAWSModule automatically adds the EC2 options, you don't need to use ec2_argument_spec any more

psharkey replied:
https://github.com/ansible/ansible/pull/23382/commits/450fd12d0898960b65177a0d10f14765fa39d5fc removes this.

tremble commented on lib/ansible/modules/cloud/amazon/iam_user_info.py:
```diff
@@ -0,0 +1,184 @@
+
+    region, ec2_url, aws_connect_params = get_aws_connection_info(module, boto3=True)
+
+    connection = boto3_conn(module, conn_type='client', resource='iam', region=region, endpoint=ec2_url, **aws_connect_params)
```
AnsibleAWSModule provides a nice helper
```suggestion
    connection = module.client('iam')
```

psharkey replied:
https://github.com/ansible/ansible/pull/23382/commits/0fd12989d28ed8304e6cdaaa1b9437d32dc08867 changes this.

tremble commented on test/integration/targets/iam_user/tasks/main.yml:
```diff
@@ -0,0 +1,119 @@
+      region: "{{ aws_region }}"
+  block:
+
+  - name: ensure ansible user exists
```
It'd be good to add some assertions here, additionally you can cross-check that the user_id returned from iam_user matches the user_id returned by iam_user_info (just in case the search was doing something screwy)

psharkey replied:
https://github.com/ansible/ansible/pull/23382/commits/518e1f4cb5c43ce481747a143ebd103d01796cec has updates so the values returned can be validated against the user created directly.

tremble commented on test/integration/targets/iam_user/tasks/main.yml:
```diff
@@ -0,0 +1,119 @@
+    ignore_errors: yes
+    register: iam_user_info_path_group
+
+  - name: assert improper usage of parameters fails gracefully
```
I tend to drop the parameter tests right at the top, they're quick and failure means you don't need to create anything.

psharkey replied:
https://github.com/ansible/ansible/pull/23382/commits/510e22e39c839f96c157306b19f0a2563dbbe2ab moves this.

tremble commented on test/integration/targets/iam_user/tasks/main.yml:
```diff
@@ -0,0 +1,119 @@
+      state: present
+    register: iam_group
+
+  - name: get info on IAM user(s) in group
```
Might be good to add some additional tests adding a second user and making sure that

1. it doesn't show up when not added to the group
2. it does show up when added to the group
3. iam_user_info behaves properly when returning multiple values

(you'd be surprised at some of the dumb bugs I've seen over the years)

psharkey replied:
https://github.com/ansible/ansible/pull/23382/commits/9e3eda56d319ccd6dcbbc1d7b513405cd1301e10 adds some test cases.

tremble commented on lib/ansible/modules/cloud/amazon/iam_user_info.py:
```diff
@@ -0,0 +1,184 @@
+    group = module.params.get('group')
+    path = module.params.get('path')
+
+    params = dict(MaxItems=1000)
```
This could in theory catch people out.  Unfortunately I don't have any experience with boto pagination...

psharkey replied:
https://github.com/ansible/ansible/pull/23382/commits/83e179a77b828b6e2fa3e472d74ad0fdae2e5197 updates this.

tremble reviewed (changes requested):
Thanks for your contribution, I finally had chance to dig a little deeper...

One additional change needed, please update lib/ansible/config/module_defaults.yml to add iam_user_info to the AWS group (yay, using module_defaults actually caught this!)

psharkey committed:
ec2_argument_spec not needed with AnsibleAWSModule.

psharkey committed:
Switch to use helper in AnsibleAWSModule.

psharkey committed:
Add iam_user_info to the aws group.

tremble commented on lib/ansible/modules/cloud/amazon/iam_user_info.py:
```diff
@@ -175,7 +172,7 @@ def main():
 
     region, ec2_url, aws_connect_params = get_aws_connection_info(module, boto3=True)
```
You can delete this row too now

psharkey replied:
https://github.com/ansible/ansible/pull/23382/commits/450fd12d0898960b65177a0d10f14765fa39d5fc removes this.

willthames replied:
It doesn't

psharkey replied:
Oops.. https://github.com/ansible/ansible/pull/23382/commits/34cc91a8d2b54d6436bda2c744f777fdc9600ebc corrects this now.

psharkey committed:
Add support for pagination and backoff.

psharkey committed:
Check improper parameter usage first.

psharkey committed:
Adding test cases for multiple users.

willthames commented on lib/ansible/modules/cloud/amazon/iam_user_info.py:
```diff
@@ -0,0 +1,187 @@
+'''
+
+from ansible.module_utils.aws.core import AnsibleAWSModule
+from ansible.module_utils.ec2 import boto3_conn, camel_dict_to_snake_dict, ec2_argument_spec, get_aws_connection_info, AWSRetry
```
This list of imports can be reduced once current line 179 is removed

psharkey replied:
https://github.com/ansible/ansible/pull/23382/commits/1db99733934e7f3f0fbcf911ceb6a9a1c8240c56 removes the imports not needed anymore.

willthames commented on lib/ansible/modules/cloud/amazon/iam_user_info.py:
```diff
@@ -0,0 +1,187 @@
+            params['UserName'] = name
+        try:
+            iam_users.append(connection.get_user(**params)['User'])
+        except (ClientError, ParamValidationError) as e:
```
any reason this doesn't use BotoCoreError rather than ParamValidationError?

I'm a bit worried this might not catch ProfileNotFound or permissions errors (one of the two)

tremble replied:
Nice catch, I'd assumed they were ClientErrors, just checked, they're BotoCoreErrors

psharkey replied:
https://github.com/ansible/ansible/pull/23382/commits/fdeb8d2f39154ab8e6d072a349d69726a7b6e9b8 switches this.

psharkey replied:
https://github.com/ansible/ansible/pull/23382/commits/a1dcf29f24549a6879e90ca9b8b1e7c98231718f adds some tests for this as well.

psharkey committed:
Rmoving unneeded line.

psharkey committed:
Remove unneeded imports.

psharkey committed:
Switch to catch BotoCoreError.

psharkey committed:
Adding tests for exception coverage.

willthames commented:
@psharkey thanks for fixing all my comments. I'll try and run the test suite over the next couple of days, but from a code review perspective, LGTM.

psharkey committed:
Compare user info directly with values from created user.

tremble commented on test/integration/targets/iam_user/tasks/main.yml:
```diff
@@ -0,0 +1,228 @@
+    set_fact:
+        test_iam_user: '{{ iam_user.iam_user.user }}'
+  - assert:
+      that:
```
Since these are also tests for iam_user we should at least run a basic test here:

```suggestion
      that:
        - 'iam_user is changed'
```

Since your PR isn't for iam_user I don't think it's worth holding it up on thorough testing of iam_user, especially with an iam_user boto3 PR in flight.

willthames commented:
Thanks so much for this @psharkey, this was an epic one! And thanks too to @tremble for some really insightful reviewing

psharkey commented:
@tremble @willthames Thank you for taking the time to review and getting this completed. The integration test framework that's in place now makes is so much easier to refactor and add tests.

tremble commented:
@psharkey Yup, it's a shame that the test suites are currently a little spotty, but retrofitting tests is never a popular task.
